src/metabase/models/setting.clj

```diff
@@ -148,18 +149,28 @@
   (setting-name [this]
     (name this)))
 
+(defn- munge-setting-name
+  "Munge names so that they are legal for bash. Only allows for alphanumeric characters,  underscores, and hyphens."
+  [setting-nm]
+  (str/replace (name setting-nm) #"[^a-zA-Z0-9_-]*" ""))
+
 (defn- env-var-name
   "Get the env var corresponding to `setting-definition-or-name`.
    (This is used primarily for documentation purposes)."
   ^String [setting-definition-or-name]
-  (str "MB_" (str/upper-case (str/replace (setting-name setting-definition-or-name) "-" "_"))))
+  (str "MB_" (-> (setting-name setting-definition-or-name)
+                 munge-setting-name
+                 (str/replace "-" "_")
+                 str/upper-case)))
 
 (defn env-var-value
   "Get the value of `setting-definition-or-name` from the corresponding env var, if any.
-   The name of the Setting is converted to uppercase and dashes to underscores;
-   for example, a setting named `default-domain` can be set with the env var `MB_DEFAULT_DOMAIN`."
+   The name of the Setting is converted to uppercase and dashes to underscores; for example, a setting named
+  `default-domain` can be set with the env var `MB_DEFAULT_DOMAIN`. Note that this strips out characters that are not
+  legal for shells. Setting `foo-bar?` will expect to find the key `:mb-foo-bar` which will be sourced from the
+  environment variable `MB_FOO_BAR`."
   ^String [setting-definition-or-name]
-  (let [v (env/env (keyword (str "mb-" (setting-name setting-definition-or-name))))]
+  (let [v (env/env (keyword (str "mb-" (munge-setting-name (setting-name setting-definition-or-name)))))]
```

Comment:
this is the one-line fix here where we get the setting from the environment using a munged name. Due to the different codepaths can't reuse the `:munged-name` key value in the setting definition so have to munge again.